File: lab4/g.py
import sys
sys.setrecursionlimit(300000)

class Node:
    __slots__ = ("data", "left", "right")
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class BST:
    def __init__(self):
        self.root = None
        self.diameter = 0

    def insert(self, node, val):
        if node is None:
            return Node(val)
        if val < node.data:
            node.left = self.insert(node.left, val)
        elif val > node.data:
            node.right = self.insert(node.right, val)
        return node

    def insert_value(self, val):
        self.root = self.insert(self.root, val)

    def depth(self, node):
        if node is None:
            return 0

        left_depth = self.depth(node.left)
        right_depth = self.depth(node.right)

        # обновляем диаметр (считаем рёбра)
        self.diameter = max(self.diameter, left_depth + right_depth)

        return 1 + max(left_depth, right_depth)

    def get_diameter(self):
        self.diameter = 0
        self.depth(self.root)
        return self.diameter + 1 

def main():
    data = sys.stdin.read().strip().split()
    n = int(data[0])
    arr = list(map(int, data[1:]))

    tree = BST()
    for x in arr:
        tree.root = tree.insert(tree.root, x)

    print(tree.get_diameter())

if __name__ == "__main__":
    main()

File: lab4/g.cpp
#include <iostream>
using namespace std;

class Node
{
public:
    int data;
    Node *left;
    Node *right;

    Node(int data)
    {
        this->data = data;
        left = NULL;
        right = NULL;
    }
};

class BST
{
public:
    Node *root;
    int diameter;

    BST()
    {
        root = NULL;
        diameter = 0;
    }

    Node *insert(Node *node, int val)
    {
        if (node == NULL)
        {
            return new Node(val);
        }
        if (val < node->data)
        {
            node->left = insert(node->left, val);
        }
        else if (val > node->data)
        {
            node->right = insert(node->right, val);
        }
        return node;
    }

    void insertValue(int val)
    {
        root = insert(root, val);
    }

    int depth(Node *node)
    {
        if (node == NULL)
            return 0;

        int leftDepth = depth(node->left);
        int rightDepth = depth(node->right);

        diameter = max(diameter, leftDepth + rightDepth);

        return 1 + max(leftDepth, rightDepth);
    }

    int getDiameter()
    {
        diameter = 0;
        depth(root);
        return diameter + 1;
    }
};

int main()
{
    int n;
    cin >> n;

    BST tree;

    for (int i = 0; i < n; i++)
    {
        int x;
        cin >> x;
        tree.root = tree.insert(tree.root, x);
    }

    cout << tree.getDiameter() << "\n";

    return 0;
}